cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -162,11 +162,26 @@ void build_knn_graph(
   // search top (k + 1) neighbors
   //
 
-  const auto top_k          = node_degree + 1;
-  uint32_t gpu_top_k        = node_degree * pq.refinement_rate;
-  gpu_top_k                 = std::min<IdxT>(std::max(gpu_top_k, top_k), dataset.extent(0));
-  const auto num_queries    = dataset.extent(0);
-  const auto max_batch_size = 1024;
+  const auto top_k       = node_degree + 1;
+  uint32_t gpu_top_k     = node_degree * pq.refinement_rate;
+  gpu_top_k              = std::min<IdxT>(std::max(gpu_top_k, top_k), dataset.extent(0));
+  const auto num_queries = dataset.extent(0);
+  rmm::device_async_resource_ref workspace_mr = raft::resource::get_workspace_resource(res);
+
+  constexpr size_t kDefaultBatchSize = 1024;
+  constexpr size_t kMaxBatchSize     = 4096;  // No more perf beyond this
+  constexpr size_t kMinBatchSize     = 128;   // Too slow if smaller
+  // Heuristic: how much of the workspace we can spare for the queries.
+  // The rest is going to be used by ivf_pq::search
+  const auto workspace_queries_bytes = raft::resource::get_workspace_free_bytes(res) / 5;
+  auto max_batch_size =
+    std::min<size_t>(workspace_queries_bytes / sizeof(DataT) / dataset.extent(1), kMaxBatchSize);
+  // Heuristic: if the workspace is too small for a decent batch size, switch to use the large
+  // resource with a default batch size.
+  if (max_batch_size < kMinBatchSize) {
```

Comment:
If `max_batch_size = 128`, then we would be running with that batch size using workspace memory. Is that batch size already large enough that it is faster than keeping `defaultBatchSize` and using `large_workspace_resource`?

Reply:
Hard to say at this point. I did some benchmarking back in February hence the value, but probably we should test it more. I'd suggest we can do more thorough benchmarking later and update the constants in a separate PR.